doc/architectural/background-concepts.md

```diff
@@ -1,32 +1,824 @@
+
 \subsection SAT_section SAT and SMT
 
-To be documented.
+- Goal: Solve model-finding problem.
+- Answers: Here is a model, no model possible
+- Complexity issues
+- SAT and SMT as extremely advanced black boxes that solve
+  a good part of problems efficiently. Point out MiniSat,
+  Z3 as good examples that we use.
+- Pointer to literature: CDCL, Lazy/Eager instantiation
+
+From our wiki:
+In a Boolean formula, variables can only be assigned values from the set
+{true, false}, and they are connected using and, or and not operators.
+The Boolean satisfiability problem is to find a valuation of the variables
+of a given formula so that the formula evaluates to true, or to prove that
+there is no such valuation. SAT is an NP-complete problem, meaning
+a given solution can be always verified in an amount of time that is polynomial
+in terms of the size of the input formula, and
+there is no known algorithm that computes a solution to any given formula
+in polynomial time (and the existence of such an algorithm would prove P = NP).
 
 \subsection static_analysis_section Static analysis
 
-To be documented.
+While BMC analyses the program by transforming everything to logic
+formulas and, essentially, running the program on sets of concrete
+states, another approach to learn about a program is based on the idea
+of interpreting an abstract version of the program. This is known
+as **abstract interpretation**. Abstract interpretation is one of the
+main methods in the area of **static analysis**.
+
+The key idea is that instead of looking at concrete program states
+(e.g., ''variable x contains value 1''), we look at some
+sufficiently-precise abstraction (e.g., ''variable x is odd'', or
+''variable x is positive''), and perform interpretation of the program
+using such abstract values. Coming back to our running example, we wish
+to prove that the factorial function never returns 0.
+
+The first ingredient we need for abstract interpretation is the
+so-called **abstract domain**. An abstract domain is a set $D$ (or, if you
+prefer, a data type) with the following properties:
+- There is a function merge that takes two elements of $D$ and returns
+  an element of $D$. This function is associative (merge(x, merge(y,z))
+  = merge(merge(x,y), z)), commutative (merge(x,y) = merge(y,x)) and
+  idempotent (merge(x,x) = x).
+- There is an element bottom of $D$ such that merge(x, bottom) = x.
+
+Algebraically speaking, $D$ needs to be a semi-lattice.
+For our example, we use the following domain:
+- D contains the elements "bottom" (nothing is known),
+  "equals 0", "not 0" and "could be 0".
+- merge is defined as follows:
+  merge(bottom, x) = x
+  merge("could be 0", x) = "could be 0"
+  merge(x,x) = x
+  merge("equals 0", "not 0") = "could be 0"
+- bottom is bottom, obviously.
+It is easy but tedious to check that all conditions hold.
+
+The domain allows us to express what we know about a given variable or
```

Comment:
First intuition, then formalization.

Put this paragraph first. Use it to justify why you need the merge operator in the formalization. Then present the lattice.